[LLVM] Add `llvm.experimental.vector.compress` intrinsic

lawben opened the pull request:
This PR adds a new vector intrinsic `@llvm.experimental.vector.compress` to "compress" data within a vector based on a selection mask, i.e., it moves all selected values (i.e., where `mask[i] == 1`) to consecutive lanes in the result vector. A `passthru` vector can be provided, from which remaining lanes are filled.

The main reason for this is that the existing `@llvm.masked.compressstore` has very strong constraints in that it can only write values that were selected, resulting in guard branches for all targets except AVX-512 (and even there the AMD implementation is _very_ slow). More instruction sets support "compress" logic, but only within registers. So to store the values, an additional store is needed. But this combination is likely significantly faster on many target as it avoids branches.

In follow up PRs, my plan is to add target-specific lowerings for x86, SVE, and possibly RISCV. I also want to combine this with a store instruction, as this is probably a common case and we can avoid some memory writes in that case.

See [discussion in forum](https://discourse.llvm.org/t/new-intrinsic-for-masked-vector-compress-without-store/78663) for initial discussion on the design.

lawben committed:
Add initial code for @llvm.masked.compress intrinsics

lawben committed:
Remove requirements for legal types

lawben committed:
Add tests for AArch64

lawben committed:
Add floating point test

lawben committed:
Add documentation

lawben committed:
Fix formatting

arsenm reviewed (commented):
Missing globalisel implementation

efriedma-quic commented:
> Is there a place to put the duplicated code between the DagTypeLegalizer and the VectorLegalizer?

TargetLowering has a bunch of expand* routines.

> SplitVecOperand or WidenResultVector. Maybe someone could point me in the right direction here, as none of my test cases fall into these paths.

Suggested types to test with:

- `<3 x i32>`.
- `<vscale x 4 x double>`

It should be possible to exercise all forms of result legalization.  Some forms of operand legalization might be impossible to hit: we legalize results before operands, so the possible operand types are restricted.

fhahn commented:
also cc @nikolaypanchenko for the RISCV side

lawben committed:
Fix references in docs

lawben committed:
Add widen for vector type legalization

lawben committed:
Put expand logic in TargerLowering to avoid code duplication.

lawben committed:
Fix formatting

lawben commented:
@arsenm I'm struggling a bit with the GlobalISel implementation, as I don't have experience with it. I've created a `GenericOpcode` for it in `GenericOpcodes.td` as follows:

```
// Generic masked compress.
def G_MCOMPRESS: GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins type1:$vec, type2:$mask);
  let hasSideEffects = false;
}
```

and I've added 

```
def : GINodeEquiv<G_MCOMPRESS, masked_compress>;
``` 

to `SelectionDAGCompat.td` by following some examples/tutorials. 

However, I cannot get any further with this, as other instructions "fail" in GlobalISel before I even get to my masked compress. As I understand, AArch64 has one of the more sophisticated GISel implementations, but even with that, I'm getting errors like:
```
LLVM ERROR: unable to legalize instruction: %1:_(<4 x s1>) = G_TRUNC %2:_(<4 x s16>)
``` 

This truncation is unavoidable in my case, as I need a `<N  x s1>` mask, which is not legal in itself.

Also, my current SelectionDAG-based implementation has a generic expand path in `TargetLowering.cpp` for all tagets that don't support it natively. I cannot seems to find a target-independent code path for GIsel in which I could add some logic to expand this.

Could you maybe point me in the right direction to continue here? I'm not really sure how to continue / what to actually implement for GIsel.

arsenm commented:
> def : GINodeEquiv<G_MCOMPRESS, masked_compress>;

This will help you with step 3 (select), but some work will be needed to define some legalizer rules (and regbankselect for the legal cases) 



> 
> However, I cannot get any further with this, as other instructions "fail" in GlobalISel before I even get to my masked compress. As I understand, AArch64 has one of the more sophisticated GISel implementations, but even with that, I'm getting errors like:
> 
> ```
> LLVM ERROR: unable to legalize instruction: %1:_(<4 x s1>) = G_TRUNC %2:_(<4 x s16>)
> ```
> 

AArch64 is fairly incomplete on vectors last I checked. It's still possible to write a MIR only test that shows your expansion without handling everything 

> This truncation is unavoidable in my case, as I need a `<N x s1>` mask, which is not legal in itself.
> 
> Also, my current SelectionDAG-based implementation has a generic expand path in `TargetLowering.cpp` for all tagets that don't support it natively. I cannot seems to find a target-independent code path for GIsel in which I could add some logic to expand this.

This would be in LegalizerHelper, conditionally called depending on the legalizer rules defined by the target (e.g. https://github.com/llvm/llvm-project/blob/c675a58edec6d1a876a0d0e7d261f74764855b38/llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp#L920)

topperc commented on llvm/docs/LangRef.rst:
```diff
@@ -24975,6 +24975,85 @@ The '``llvm.masked.compressstore``' intrinsic is designed for compressing data i
+Semantics:
+""""""""""
+
+The '``llvm.masked.compress``' intrinsic is designed for compressing data within a vector, i.e., ideally within a register.
```
This description doesn't describe semantics. It explains the utility.

topperc commented on llvm/docs/LangRef.rst:
```diff
@@ -24975,6 +24975,85 @@ The '``llvm.masked.compressstore``' intrinsic is designed for compressing data i
+Selects elements from input vector '``value``' according to the '``mask``'.
+All selected elements are written into adjacent lanes in the result vector, from lower to higher.
+The mask holds a bit for each vector lane, and is used to select elements to be kept.
+The number of valid lanes is equal to the number of active bits in the mask.
```
I think this is trying to say that the upper elements are undefined or poison?

lawben replied:
yes, i've explicitly added this now.

topperc commented on llvm/docs/LangRef.rst:
```diff
@@ -24975,6 +24975,85 @@ The '``llvm.masked.compressstore``' intrinsic is designed for compressing data i
+    %vec = load <4 x i32>, ptr %inPtr
+    %mask = load <4 x i1>, ptr %maskPtr
+    %compressed = call <4 x i32> @llvm.masked.compress.v4i32(<4 x i32> %vec, <4 x i1> %mask)
+    store <4 x i32> %compressed, ptr %outPtr
```
If any bits in the mask are 0, this stores garbage to those elements. If this occurs on the last iteration or the mask is all zeros for all future iterations, those elements will never get overwritten.

Isn't that different than what the scalar code does? The scalar code will leave those elements untouched.

lawben replied:
you are right. I tried to keep the example similar to `compressstore` but I think that is not a good choice. I've updated it and removed the `store` because that complicates things from a general semantic explanation. It is up to the user to decide how to write this, if at all.

lawben committed:
Add basic lowering of MCOMPRESS in GlobalISel

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7502,6 +7504,53 @@ LegalizerHelper::lowerShuffleVector(MachineInstr &MI) {
+LegalizerHelper::lowerMCOMPRESS(llvm::MachineInstr &MI) {
+  auto [Dst, DstTy, Vec, VecTy, Mask, MaskTy] = MI.getFirst3RegLLTs();
+
+  if (VecTy.isScalableVector())
+    report_fatal_error(
+        "Lowering masked_compress for scalable vectors is undefined.");
```
I'd expect this to be caught in the IR verifier, and then you don't need to worry about it

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7502,6 +7504,53 @@ LegalizerHelper::lowerShuffleVector(MachineInstr &MI) {
+  MIRBuilder.buildLoad(Dst, StackPtr, PtrInfo, VecAlign);
+
+  MI.eraseFromParent();
+  // TODO: This is not true! We don't know if the input vector type is legal.
+  //       Find out how to assert this!
```
Delete this comment. It doesn't matter if the final results are legal, it just needs to report that it did make a change as requested.

lawben committed:
Add basic AArch64 MIR test

lawben committed:
Address PR comments

RKSimon commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1294,6 +1294,11 @@ enum NodeType {
+  // Masked compress - consecutively place vector elements based on mask
+  // e.g., vec = {A, B, C, D} and mask = 1010
+  //         --> {A, C, ?, ?} where ? is undefined
+  MCOMPRESS,
```
Please document the operand order

lawben replied:
Done. Is this enough? I've followed the examples of the other instructions.

arsenm replied:
Expand the name to exactly match the intrinsic name

RKSimon reviewed (commented):
Maybe worth adding ISD::MCOMPRESS to VerifySDNode to ensure the types are correct?

lawben committed:
Update docs according to PR comments

lawben commented:
> Maybe worth adding ISD::MCOMPRESS to VerifySDNode to ensure the types are correct?

@RKSimon that's a good idea. As `VerifySDNode` is still quite new, I'm not sure if this is intended behavior or not, but when implementing `MCOMPRESS` in it, I get crashes in my program. Specifically, in `SelectionDAG::getNode()`, there is a step that tries to `FoldConstantArithmetic`, which creates an `MCOMPRESS` node with scalar input. As this is not valid, my verify code crashes. I assume this will happen for many other nodes that require vector inputs.

I could add a check at the beginning of `FoldConstantArithmetic` and bail, like `CONCAT_VECTORS` 
https://github.com/llvm/llvm-project/blob/84aee95124549c5d13e22053af254e3fcc02bc84/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L6279-L6283

but that feels like the wrong place/abstraction to do this at. Do you have any suggestions where to put logic to avoid this? In my opinion, the crash in `VerifySDNode` is correct here and the scalar path should never be tried, so we probably want to avoid it.

lukel97 commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -2362,6 +2362,11 @@ def int_masked_compressstore:
 
+def int_masked_compress:
+    DefaultAttrsIntrinsic<[llvm_anyvector_ty],
+              [llvm_anyvector_ty, LLVMScalarOrSameVectorWidth<0, llvm_i1_ty>],
```
I think we can enforce that the two input and output vector types match, something like
```suggestion
              [LLVMMatchType<0>, LLVMScalarOrSameVectorWidth<0, llvm_i1_ty>],
```

lawben replied:
Good catch.

lukel97 commented on llvm/docs/LangRef.rst:
```diff
@@ -24975,6 +24975,72 @@ The '``llvm.masked.compressstore``' intrinsic is designed for compressing data i
 
 Other targets may support this intrinsic differently, for example, by lowering it into a sequence of branches that guard scalar store operations.
 
+Masked Vector Compress Intrinsic
```
This might be bikeshedding, but would it make more sense to put this in the llvm.vector.* family of intrinsics? Since otherwise this would be the only masked intrinsic that doesn't read or write memory

lawben replied:
Sounds good. I've moved it to the `vector.*` section.

lawben replied:
After re-reading this, I think I misunderstood your comments, sorry. You were advocating for calling the the intrinsic `@llvm.vector.compress` instead of `@llvm.masked.compress` and not just moving this in the docs? @lukel97 @nikolaypanchenko Could you please confirm this?

lukel97 replied:
Yeah that's what I originally had in mind, but I'm not strongly opinionated so I'll leave this up to you/other reviewers. This PR looks like it's making good progress and I don't want to block it :)

lawben replied:
The more I think about it, the more I think you are right about the naming. The "mask" is an essential part of the operation, as @nikolaypanchenko points out, so it does not make sense to use it to distinguish between a masked and non-masked compress. 

Pulling in @efriedma-quic @RKSimon for an opinion on naming and putting it in the experimental namespace.

If I change it, I'd probably change it to `@llvm.experimental.vector.compress` as suggested by @lukel97 , as this covers both the `vector` and `experimental` (see [other comment](https://github.com/llvm/llvm-project/pull/92289/#discussion_r1646842999) for discussion on this) for now.

efriedma-quic commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11226,6 +11226,53 @@ SDValue TargetLowering::expandVectorSplice(SDNode *Node,
+  EVT ScalarVT = VecVT.getScalarType();
+  EVT MaskScalarVT = Mask.getValueType().getScalarType();
+
+  if (VecVT.isScalableVector())
+    report_fatal_error(
+        "Expanding masked_compress for scalable vectors is undefined.");
```
masked_compress is perfectly well-defined for scalable vectors... the problem is that we can't emit a loop in SelectionDAG.  We can expand it in IR if it turns out to be relevant.

lawben replied:
I've changed the error message to be a bit more clear. My understanding is that we need target-specific custom lowering for SVE and RISCV.

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2335,6 +2347,19 @@ SDValue DAGTypeLegalizer::PromoteIntOp_MSCATTER(MaskedScatterSDNode *N,
+  SDValue Mask = N->getOperand(1);
+  EVT VT = Vec.getValueType();
+
+  if (OpNo == 0)
```
Because the result type and first operand type match, this will never be called with OpNo == 0.

lawben replied:
Changed to `assert(OpNo == 1)`.

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -2304,6 +2307,25 @@ void DAGTypeLegalizer::SplitVecRes_Gather(MemSDNode *N, SDValue &Lo,
+  // of MCOMPRESS and all other operands can be legalized later.
+  SDValue Compressed = TLI.expandMCOMPRESS(N, DAG);
+
+  SDLoc DL(N);
```
Can we use DAG.getSplitVector here?

lawben replied:
Yes, fixed.

RKSimon commented:
If FoldConstantArithmetic is already special casing ISD::CONCAT_VECTORS I don't see why we shouldn't just special case another generic ISD as well - we could create a ISD wrapper if you prefer e.g. ISD::isVectorOnlyOp(unsigned Opcode) ?

topperc commented:
> > Maybe worth adding ISD::MCOMPRESS to VerifySDNode to ensure the types are correct?
> 
> @RKSimon that's a good idea. As `VerifySDNode` is still quite new, I'm not sure if this is intended behavior or not, but when implementing `MCOMPRESS` in it, I get crashes in my program. Specifically, in `SelectionDAG::getNode()`, there is a step that tries to `FoldConstantArithmetic`, which creates an `MCOMPRESS` node with scalar input. As this is not valid, my verify code crashes. I assume this will happen for many other nodes that require vector inputs.
> 
> I could add a check at the beginning of `FoldConstantArithmetic` and bail, like `CONCAT_VECTORS`
> 
> https://github.com/llvm/llvm-project/blob/84aee95124549c5d13e22053af254e3fcc02bc84/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L6279-L6283
> 
> but that feels like the wrong place/abstraction to do this at. Do you have any suggestions where to put logic to avoid this? In my opinion, the crash in `VerifySDNode` is correct here and the scalar path should never be tried, so we probably want to avoid it.

Why can't we check the types in getNode like we do for most nodes. This node isn't variadic so it doesn't have a different number of operands like BUILD_VECTOR. I'm not sure why BUILD_PAIR is VerifySDNode.

lawben committed:
Match result and input types of MCOMPRES

lawben committed:
Add constant folding to SelectionDAG::getNode()

lawben committed:
Address PR comments for type legalization

lawben committed:
Move masked.compress in docs

lawben commented:
> Why can't we check the types in getNode like we do for most nodes.

@topperc You are right. I've moved the code there now. I think it makes more sense there. I've also added code for constant folding, as some of the other instructions also do. 

@RKSimon I've also added a check in `FoldConstantArithmetic` to avoid `MCOMPRESS`, as this is now handled in `getNode()`.

lawben committed:
Fix bug for x86 result widening

lawben committed:
Remove zero-fill when widening vector

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7197,6 +7199,48 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
   }
+  case ISD::MCOMPRESS: {
+    EVT VecVT = N1.getValueType();
+    EVT MaskVT = N2.getValueType();
+    (void)MaskVT;
```
Nit you can use maybe_unused instead to prevent warnings
```suggestion
    [[maybe_unused]] EVT MaskVT = N2.getValueType();
```

lawben replied:
okay, changed. I thought of this but did not find any use of `[[maybe_unused]]` in this file, so I stuck with the file's convention.

lukel97 replied:
Oh sorry didn't realise that's what the rest of the file was using. We should probably mop up the other void uses now that we've migrated to C++17 (in another PR!)

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -6279,7 +6279,9 @@ SDValue SelectionDAG::FoldConstantArithmetic(unsigned Opcode, const SDLoc &DL,
-  if (Opcode >= ISD::BUILTIN_OP_END || Opcode == ISD::CONCAT_VECTORS)
+  // MCOMPRESS is not defined for scalars. We handle constants before this call.
+  if (Opcode >= ISD::BUILTIN_OP_END || Opcode == ISD::CONCAT_VECTORS ||
+      Opcode == ISD::MCOMPRESS)
```
How come we need to bail here for MCOMPRESS but not for other vector binary ops like VECTOR_DEINTERLEAVE etc?

lawben replied:
I had the same thought and did some digging. I think this is somewhat coincidental, as a few things have to happen to run into this case. I also believe there is a bug in, e.g., `VECTOR_DEINTERLAVE` that you compared to.

`FoldConstantArithmetic` on the scalar elements of a vector is only applied when...

1. the node has one or two operands, not more (disqualifies, e.g., `VECTOR_SPLICE`, `INSERT_SUBVECTOR`, as both have a third index operand)
2. all operands are vectors with the same number of elements (disquaifies, e.g., `EXTRACT_SUBVECTOR`, as the index argument is not a vector)
3. the vectors must all be `BUILD_VECTOR`, `VECTOR_SPLAT`, or `UNDEF`.
4. the op code must actually be invalid for scalar types (disqualifies, e.g., all `VECTOR_REDUCE*`)

on top of that, this crash only happens when the ISD op code has `asserts` for `isVector()` in `getNode()`. e.g., `VECTOR_DEINTERLEAVE` does not have this. so technically, we can create an invalid `VECTOR_DEINTERLEAVE` node with two scalars, which sounds like a bug to me. 

```cpp
SDValue test = getNode(ISD::VECTOR_DEINTERLEAVE, DL, MVT::i32, 
                       getConstant(0, DL, MVT::i32), 
                       getConstant(1, DL, MVT::i32));
```

This code happily creates a SelectionDAG node.

```
Creating new node: t24: i32 = vector_deinterleave Constant:i32<0>, Constant:i32<1>
```

Due to the suggestion of @RKSimon and @topperc, I added some asserts to avoid this case and I crash. 

**Option 1:** Allow scalar operands for these vector operations.
**Option 2:** Make `FoldConstantArithmetic` more aware of vector-only functions, and skip them.

What are your thougths on this?

lukel97 replied:
Yikes, thanks for the explanation. I don't have much of an opinion on how to handle the other ops but I think what you have here for MCOMPRESS is perfectly reasonable. We shouldn't worry about the others in this PR

lukel97 commented on llvm/docs/LangRef.rst:
```diff
@@ -19197,6 +19197,71 @@ the follow sequence of operations:
+
+.. _int_masked_compress:
+
+'``llvm.masked.compress.*``' Intrinsics
```
Should we mark this as experimental? Something like `llvm.experimental.vector.compress`?

lawben replied:
I do not know what the convention is for adding this in experimental or not. Would the flow be to have it experimental for a release or two and then make it "normal"? I don't have an opinion on this, so I'm happy to follow whatever the recommended way is.

lukel97 replied:
Sorry for the delay, yeah it looks like they're removed after about a release or two, see bfc0317153dca75137fba00b5c28758d6f720963. I think it might not be a bad idea just in case we want to change things after this PR. 

Also on the topic of naming, if this in the vector family of intrinsics should the name be updated to reflect that? E.g. `@llvm.vector.blah`

lawben committed:
Use [[maybe_unused]] for asserts

lawben commented:
I've added a [placeholder PR](https://github.com/lawben/llvm-project/pull/1) in my fork for SVE lowering on AArch64, so you can see how I will continue.  Next steps would be custom lowering for x86 and RISCV. And I would like to do a target-independent compress+store combine, to avoid the temporary stack stores of we just load+store again.

Question to all: What is missing for this PR to be merged? I guess the GlobalISel stuff is still open (@arsenm), but I need some input there. Other than that, what do I still need to implement/do?

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7197,6 +7200,47 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+           "Vector and mask must have same number of elements.");
+
+    APInt SplatVal;
+    if (ISD::isConstantSplatVector(N2.getNode(), SplatVal))
```
I'm not sure it makes sense to do a lot of folding in getNode. It would only trigger if the mcompress is created with a constant vector. It's more likely the mask becomes a constant as part of DAG combine which this code won't catch. So I think its better to do this in DAGCominer.

lawben replied:
done.

topperc commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1132,6 +1132,10 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .scalarize(1)
       .lower();
 
+  // TOOD: FIX ThiS!
```
TOOD -> TODO

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7197,6 +7200,47 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+    assert(VT == VecVT && "Vector and result type don't match.");
+    assert(VecVT.isVector() && MaskVT.isVector() &&
+           "Both inputs must be vectors.");
+    assert(VecVT.isScalableVector() == MaskVT.isScalableVector() &&
```
This should be covered by the getVectorElementCount() check. Element count includes whether the vector is scalable.

lawben committed:
Move constant folding to DAGCombiner

lawben committed:
Change TODO for AArch64 GlobalISel

lawben committed:
Remove wrong ISA from docs

lawben commented:
@arsenm could you let me know how to proceed with the GlobalISel stuff on Arch64? I can also move this to the follow-up PR ([temporary version here](https://github.com/lawben/llvm-project/pull/1)), as that covers AArch64 explicitly anyway.

arsenm commented on llvm/include/llvm/Target/GenericOpcodes.td:
```diff
@@ -1500,6 +1500,13 @@ def G_SPLAT_VECTOR: GenericInstruction {
 }
 
+// Generic masked compress.
+def G_MCOMPRESS: GenericInstruction {
```
Should match the IR operation name and spill out the masked_compress

lawben replied:
done. renamed to `G_MASKED_COMPRESS`

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7502,6 +7504,47 @@ LegalizerHelper::lowerShuffleVector(MachineInstr &MI) {
+    }
+  }
+
+  Align VecAlign = getStackTemporaryAlignment(VecTy);
```
This should not query the alignment for the type again. It would be better to query the exact alignment from the frame index that was previously created

lawben replied:
done

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11226,6 +11226,53 @@ SDValue TargetLowering::expandVectorSplice(SDNode *Node,
+  SDValue Chain = DAG.getEntryNode();
+  SDValue OutPos = DAG.getConstant(0, DL, MVT::i32);
+
+  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
```
This is this, drop all the TLI uses

lawben replied:
done.

arsenm commented:
> @arsenm could you let me know how to proceed with the GlobalISel stuff on Arch64? I can also move this to the follow-up PR ([temporary version here](https://github.com/lawben/llvm-project/pull/1)), as that covers AArch64 explicitly anyway.

Separate PR for the legal case makes sense

lawben committed:
Rename MCOMPRESS to MASKED_COMPRESS

lawben committed:
Get stack alignment for load in GlobalISel

lawben committed:
Fix formatting

lawben committed:
Replace use of TLI inside of TargetLowering

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7502,6 +7504,48 @@ LegalizerHelper::lowerShuffleVector(MachineInstr &MI) {
+
+  MachineFrameInfo &MFI = MI.getMF()->getFrameInfo();
+  MIRBuilder.buildLoad(Dst, StackPtr, PtrInfo,
+                       MFI.getObjectAlign(PtrInfo.StackID));
```
The StackID is not the frame index, This just happens to work because the default stack ID is 0 and the one object you have is also a 0. The ABI for this is unfortunately awkward.

The API here could probably use work. createStackTemporary returns the G_FRAME_INDEX which contains the index you really want. We could maybe return the instruction and the actual index as a pair from createStackTemporary?

lawben replied:
Just to validate my understanding: it is possible for `CreateStackObject` to change the alignment, so we technically have no guarantee that the alignment that we pass in is the stack's alignment that we actually use?

I've seen two other approaches to this in `LegalizerHelper`. 

1) pass the alignment to `createStackTemporary()` and then just reuse that alignment object later on (somewhat similar to my first approach but without the redundant call to get the alignment). If my understanding is correct, this could technically lead to wrong alignment if the requested alignment is larger than `MachineFrameInfo::StackAlignment`. 

2) a somewhat awkward `FIDef->getOperand(1).getIndex()` in `lowerMemset()`, which I think could easily break if something changes.
https://github.com/llvm/llvm-project/blob/d5fca0f83971c8a3b9aa4a34cb8301cd4e0bf363/llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp#L8615

I think changing the API to make it easier to get the real alignment makes sense. But this is probably a separate PR, as this would mean each caller must think about what happens when the input alignment is `!=` the frame's alignment. In light of this, I'd suggest to use `VecAlign` here, as done in approach 1) with a TODO to fix this when the API changes.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7502,6 +7504,48 @@ LegalizerHelper::lowerShuffleVector(MachineInstr &MI) {
+  LLT ValTy = VecTy.getElementType();
+  Align ValAlign = getStackTemporaryAlignment(ValTy);
+
+  Register OutPos = MIRBuilder.buildConstant(IdxTy, 0).getReg(0);
```
Can use auto and drop the .getReg(0)s throughout on OutPos

lawben replied:
done.

lawben committed:
Use vector alignement for stack load

lawben committed:
Add llvm.masked.compress to release notes

lawben committed:
Merge branch 'main' into masked_compress

lawben committed:
Merge branch 'main' into masked_compress

lawben committed:
Merge branch 'main' into masked_compress

lawben commented:
@arsenm @topperc @RKSimon @efriedma-quic I believe this PR is ready to go from my side, having addressed all comments so far. Is there something missing for approval?

RKSimon commented:
The only thing I'm still wondering about is whether there should be an initial 'background' value argument as well - currently all remaining upper elements are assumed to be undef, but I often at least need these to be zero. Creating a post-selection mask that is compatible with the compression mask can be rather annoying.

lawben commented:
In the [initial discussion](https://discourse.llvm.org/t/new-intrinsic-for-masked-vector-compress-without-store/78663/4), it sounded like zeroing out could be an issue for RISC-V (and the current scalar fallback implementation), while AVX512 and SVE (can) zero out the remaining lanes. Because of this, the poison/undef version seemed to have the most support. I guess it is a trade-off between the the performance of just "compressing" (where undef may have better performance) vs. the ergonomics of compressing with 0s in remaining lanes. 

In my opinion, the performance argument "wins" here, as it is supposed to be a faster alternative to `compressstore`. I believe the assumptions there are too strong. So going for strong guarantees again might defeat the purpose.

efriedma-quic commented:
I mean, really, the issue here is the underlying instructions have functionality to set the trailing elements that isn't getting exposed by the intrinsic, and if the elements are always poison, there isn't any easy way to get access to that functionality.  If we don't think the functionality is important, that's fine... but if it is important, we should expose it as a "passthrough" input.  Trying to pattern-match it seems unreliable.

I'm not really familiar with the algorithms that use compress, so I'm not sure how important it is to set the trailing elements.

RKSimon commented:
For reference this is what I had in mind:
```cpp
// Consecutively place selected values in a vector.
using VecT __attribute__((vector_size(N))) = int;
VecT compress(VecT base, VecT vec, VecT mask) {
  VecT out = base;
  int idx = 0;
  for (int i = 0; i < N / sizeof(int); ++i) {
    out[idx] = vec[i];
    idx += static_cast<bool>(mask[i]);
  }
  return out;
}
```
If we can't do this, its not very difficult to work around (count number of active mask elements and use that to create a shuffle/select mask), but it is extra work - and the initial initialization above seems pretty trivial to add to the implementation.

lawben commented:
Mhm, yes. I think this is a tricky one. In that case, we probably want to add a passthrough value to the call. If users add a passthrough, this could either be:
  - _undef_: if the use does not care (as this is potentially more efficient)
  - _all zeros_: this should be easy to detect if it's all constants and we can use this information to select the correct instruction if there is one (e.g., in SVE or AVX512).
  - _the source vector_ or _any other vector_: then we need to maintain the input. I'm not yet sure about how to handle this ideally in the fall back implementation. For AVX512, there are explicit instructions for this, for SVE/RISC-V, we need to generate a selection mask similar to the RISC-V [code example](https://discourse.llvm.org/t/new-intrinsic-for-masked-vector-compress-without-store/78663/7?u=lawben) in the discussion thread.

If people agree that we want to expose this, then I'll add a `passthru` operand to `@llvm.masked.comress`. I think @efriedma-quic is correct and we want to expose this behavior, because doing it later is tricky. And having the option to let `passthru` be `undef` gives us the potentially faster path too.

RKSimon commented:
Thanks, I really want to apologize for not mentioning this sooner :(

lawben committed:
Add passthru vector to @llvm.masked.compress

lawben committed:
Merge branch 'main' into masked_compress

lawben committed:
Add passthru to DAGCombiner

lawben committed:
Update LangRef with passthru

lawben committed:
Add passthru to GlobalISel

lawben committed:
Merge branch 'main' into masked_compress

lawben committed:
Fix formatting

lawben committed:
Fix GlobalISel test

lawben committed:
Update comment on MASKED_COMPRESS opcode

lawben commented:
@RKSimon I think this now covers the `passthru` options listed above. The "undef" path is unchanged, the "all zeros" path is covered by checking for a splat vector, and the "other" case is the `HasPassthru` path in `TargetLowering::expandMASKED_COMPRESS()`. Implemented and tested in both SelectionDAG and GlobalISel. GlobalIsel will receive a bit more attention in the follow-up PR for AArch64, where I explicitly handle il-/legal cases.

RKSimon commented on llvm/docs/LangRef.rst:
```diff
@@ -19230,6 +19230,78 @@ the follow sequence of operations:
+        idx += static_cast<bool>(mask[i]);
+      }
+      for (int i = idx; i < N / sizeof(int); ++i) {
+        out[idx] = passthru[i - idx];
```
Does anyone need to use `passthru[i - idx]` instead of `passthru[i]`?

lawben replied:
... no idea why I mixed this up when writing the docs. Good catch. The implementation uses `passthru[i]`.

lawben committed:
Fix docs

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12091,6 +12093,57 @@ SDValue DAGCombiner::visitVP_STRIDED_STORE(SDNode *N) {
+
+  APInt SplatVal;
+  if (ISD::isConstantSplatVector(Mask.getNode(), SplatVal))
+    return SplatVal.isAllOnes()
```
Should we enforce that the contents of Mask match getBooleanContents ? And should we be flipping this logic and check for a zero or non-zero value?

lawben replied:
Good question. Now that you mentioned it, I think we should enforce getBooleanContents, because then we are on the safe side.

As the combiner may run after some initial type legalization, the mask may not be a vector of `i1` anymore. but as it started off as one, some other step must have converted the mask via the appropriate conversion.

Just checking for zero would then also be wrong, because there are configurations in which only bit 0 counts, so any leading garbage != 0 would lead to a wrong result.

i've implemented this logic now. Please check if this is better now.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12091,6 +12093,57 @@ SDValue DAGCombiner::visitVP_STRIDED_STORE(SDNode *N) {
+      SDValue Val =
+          HasPassthru
+              ? DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, ScalarVT, Passthru,
+                            DAG.getVectorIdxConstant(Rest - NumSelected, DL))
```
getVectorIdxConstant(Rest) ?

lawben replied:
fixed.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11336,6 +11336,105 @@ SDValue TargetLowering::expandVectorSplice(SDNode *Node,
+        DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, MaskScalarVT, Mask, Idx);
+    MaskI = DAG.getNode(ISD::TRUNCATE, DL, MVT::i1, MaskI);
+    MaskI = DAG.getNode(ISD::ZERO_EXTEND, DL, MVT::i32, MaskI);
+    OutPos = DAG.getNode(ISD::ADD, DL, MVT::i32, OutPos, MaskI);
```
Might not want to hard code to i32 (e.g. RISCV64 doesn't like them) - better off use getVectorIdxConstant/getVectorIdxTy ?

lawben replied:
done.

lawben committed:
Address PR comments

lawben committed:
Merge branch 'main' into masked_compress

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12078,6 +12080,68 @@ SDValue DAGCombiner::visitVP_STRIDED_STORE(SDNode *N) {
+      }
+    }();
+
+    return HasTrueBoolContent ? Vec
```
I think you can use TLI.isConstTrueVal for this

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12078,6 +12080,68 @@ SDValue DAGCombiner::visitVP_STRIDED_STORE(SDNode *N) {
+        continue;
+
+      ConstantSDNode *CMaskI = cast<ConstantSDNode>(MaskI);
+      if (CMaskI->isAllOnes()) {
```
`TLI.isConstTrueVal(MaskI)` ?

lawben committed:
Use isConstTrueVal

RKSimon reviewed (commented):
Just one more minor, does anyone else have any comments? I haven't looked much at the GISel code

lawben committed:
Remove redundant undef

efriedma-quic commented on llvm/docs/LangRef.rst:
```diff
@@ -19234,6 +19234,78 @@ the follow sequence of operations:
+The mask holds an entry for each vector lane, and is used to select elements to be kept.
+If ``passthru`` is undefined, the number of valid lanes is equal to the number of ``true`` entries in the mask, i.e., all lanes >= number-of-selected-values are undefined.
+If a ``passthru`` vector is given, all remaining lanes are filled with the corresponding lane's value from ``passthru``.
+The main difference to :ref:`llvm.masked.compressstore <int_compressstore>` is that the we do not need to guard against memory access for unselected lanes.
```
Please wrap lines at 80 columns

Probably worth explicitly stating what happens if the mask has poison/undef bits.  (And make sure the expansion freezes store indexes as necessary to implement those semantics.)

lawben replied:
Good point, I've added this in the docs. I would interpret `undef` as "false" (this how I implemented it), because we cannot access that element and the caller says "I don't care about this value". I guess this does not need any freezing stores, as we handle this deterministically.

efriedma-quic replied:
That doesn't work.  There's no way at runtime to compute whether an input is undef/poison at runtime.  (You can technically special-case a literal undef/poison, but that doesn't really do anything useful.)

lawben replied:
Ah, okay. I think my understanding of when a freeze is needed was wrong. In this case, I'm not sure we can specify one or the other behavior if a mask element is undef/poison. Depending on the code path, we perform a vec_reduce to get a passthru value or not. So our result depends on how vec_reduce behaves with undef/poison in addition to how our extracted mask element if frozen, at which point this probably gets weird to handle consistently.

Before changing the docs again: Is it valid for us to say "If the mask is undef or contains undef elements, the entire result is undef"? That would give us a bit of wiggle room in the implementation.

efriedma-quic replied:
If a mask element is poison, I think it's fine if the whole returned vector is poison.  But preferably not undefined behavior; undefined behavior means we can't hoist, which makes everything more complicated, so I'd rather not deal with that.  (I guess you could try to more narrowly state that elements after the poisoned mask bit are poison, but I don't think that's actually useful.)

If a mask element is undef... I guess we could similarly say the whole returned vector is undef?  Not sure how much it matters exactly what rule we pick here.

lukel97 replied:
What does `@llvm.masked.compressstore` currently do with undef or poison mask elements? It would be nice if they were consistent

lawben replied:
@lukel97  I just played around a bit with `@llvm.masked.compressstore` and immediate values. It is does not seem very consistent (https://godbolt.org/z/xjeY5z9bb). Sometimes and `undef` value is treated as true, sometimes as false. I'm not sure what the takeaway from this is though.

@efriedma-quic Sorry, but I'm a bit lost in your distinction here. Could you maybe suggest some wording on how to treat this? I think we basically want to say that if a mask is/contains poison/undef bits, we don't care about the result vector. I just don't know how to express this with the correct LLVM terminology.

efriedma-quic replied:
Strawman wording: "if any element of the mask is `poison`, all elements of the result are `poison`.  Otherwise, if any element of the mask is `undef`, all elements of the result are `undef`."

This mostly only affects optimizations you haven't implemented yet... but probably the code to expand llvm.masked.compress to stores needs to freeze the mask, so we don't do a store with a poisoned pointer operand.

efriedma-quic commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -5720,6 +5737,23 @@ SDValue DAGTypeLegalizer::WidenVecRes_VP_STRIDED_LOAD(VPStridedLoadSDNode *N) {
+                                    WideVecVT.getVectorNumElements());
+
+  SDValue WideVec = ModifyToType(Vec, WideVecVT);
+  SDValue WideMask = ModifyToType(Mask, WideMaskVT);
```
Do you need to pass FillWithZeros=true for the mask?

lawben replied:
In my understanding, we don't. If we have a `<3 x i32>` but need a `<4 x i32>`, the "new" fourth element is only relevant for the register but not to the user. The fourth `Vec` and `Passthru` values are both `undef`, so it does not really matter if the mask selects them or not. Please correct me if I'm wrong.

efriedma-quic replied:
Suppose you have a mask of `<false, false, true>`, and a passthru.  If you widen the mask to `<false, false, true, true>`, the second element of the output is taken from the input instead of the passthru.  (If the passthru is undef, it doesn't matter.)

efriedma-quic commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -12078,6 +12080,54 @@ SDValue DAGCombiner::visitVP_STRIDED_STORE(SDNode *N) {
+  if (ISD::isConstantSplatVector(Mask.getNode(), SplatVal))
+    return TLI.isConstTrueVal(Mask) ? Vec : Passthru;
+
+  if (Vec.isUndef() || Mask.isUndef())
```
Should this be `if ((Vec.isUndef() && Passthru.isUndef()) || Mask.isUndef())`?

lawben replied:
done.

RKSimon replied:
Can we just do this?
```c
if (Vec.isUndef() || Mask.isUndef())
  return Passthru;
```

lawben replied:
Yes, I actually like that better. This is also more consistent with the semantics of "undef bits in the mask are false", as a completely undef mask is then also completely false.

lawben committed:
Address PR comments

lawben committed:
Return passthru for undef mask or vec

RKSimon reviewed (commented):
A couple of minors

lawben committed:
Address PR comments

lawben committed:
Rename masked.compress to experimental.vector.compress

lawben committed:
Add freeze to mask extract for poison/undef entries

lawben commented:
@lukel97 I've renamed the intrinsic to `llvm.experimental.vector.compress`.
@efriedma-quic I've taken your wording and added it to the docs. I think it described the semantics clearly, so I don't see any reason to change it. I've also frozen the mask entries.

lawben committed:
Fix docs

lukel97 reviewed (commented):
The renaming looks good, thanks!

lawben committed:
Fix docs

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11336,6 +11336,108 @@ SDValue TargetLowering::expandVectorSplice(SDNode *Node,
+    Popcount = DAG.getNode(ISD::ZERO_EXTEND, DL,
+                           MaskVT.changeVectorElementType(ScalarVT), Popcount);
+    Popcount = DAG.getNode(ISD::VECREDUCE_ADD, DL, ScalarVT, Popcount);
+    SDValue LastElmtPtr =
+        getVectorElementPointer(DAG, StackPtr, VecVT, Popcount);
+    LastWriteVal = DAG.getLoad(
+        ScalarVT, DL, Chain, LastElmtPtr,
+        MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()));
+    Chain = LastWriteVal.getValue(1);
```
Can we just extract LastWriteVal from the passthru directly with extract_vector_elt?

lawben replied:
We technically could, but at least on AArch64 and x86, this boils down to the same code. There is no instruction to extract a variable position from a vector, so the vector is stored to the stack and then loaded at position `i`. In the end, we are already doing this but can avoid the extra store of the vector generated by `EXTRACT_VECTOR_ELEMENT`.

lukel97 replied:
Ah I didn't realise that that was the case for those targets. On RISC-V it can be done without a trip to the stack via `vslidedown.vi` and `vmv.x.s`, but we can natively lower the node anyway so hopefully we won't ever need to expand it.

lukel97 commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -266,6 +266,12 @@ def SDTMaskedScatter : SDTypeProfile<0, 4, [
   SDTCisSameNumEltsAs<0, 1>, SDTCisSameNumEltsAs<0, 3>
 ]>;
 
+def SDTMaskedCompress : SDTypeProfile<1, 3, [
```
```suggestion
def SDTVectorCompress : SDTypeProfile<1, 3, [
```

lukel97 reviewed (approved):
Can't speak for the GlobalISel/AArch64 stuff but the LangRef + SelectionDAG parts LGTM, thanks! Probably best if you wait for another LGTM.

lawben committed:
Address PR comments

lawben commented:
@efriedma-quic @RKSimon is there anything else to add/do? as @lukel97 suggested, I think it would be better to have more than one LGTM here

RKSimon reviewed (approved):
LGTM

efriedma-quic reviewed (approved):
LGTM

lawben committed:
Merge branch 'main' into masked_compress

RKSimon commented:
@lawben You should be OK to commit now
